File: 132 Large repunit factors.pl
#!/usr/bin/perl

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 13 May 2017
# https://github.com/trizen

# https://projecteuler.net/problem=132

# Runtime: 5 min, 14 sec

use 5.010;
use strict;
use warnings;

use ntheory qw(prime_iterator powmod gcd);

my $n     = 1e9;
my $count = 40;
my $iter  = prime_iterator();

my $factors_sum = 0;

OUTER: while (my $p = $iter->()) {

    gcd($n, $p) == 1 or next;

    my $sum    = 0;
    my $period = 0;

    foreach my $k (0 .. $p) {
        $sum += powmod(10, $k, $p);
        $sum %= $p;
        $period += 1;

        if ($sum == 0) {
            ($n % $period == 0) ? last : next OUTER;
        }
    }

    if ($sum == 0) {
        $factors_sum += $p;
        say "[$count] $p is a prime factor -> $factors_sum";
        last if --$count <= 0;
    }
}

say $factors_sum;

File: 132 Large repunit factors.jl
#!/usr/bin/julia

# Daniel "Trizen" Șuteu
# License: GPLv3
# Date: 14 May 2017
# https://github.com/trizen

# https://projecteuler.net/problem=132

# Runtime: 6 min, 14 sec

using Primes

function p_132()

    count = 40
    n = 10^9

    p = 1
    t = 0

    while true
        p += 1

        isprime(p)     || continue
        gcd(n, p) == 1 || continue

        sum = 0
        period = 0

        for k in 0:p
            sum += powermod(10, k, p)
            sum %= p
            period += 1
            sum == 0 && break
        end

        if (sum == 0 && n % period == 0)
            t += p
            println("[$count] $p is a prime factor -> $t")
            count -= 1
            count == 0 && break
        end
    end

    return t
end

println(p_132())
